src/VariablesState.ts

```diff
@@ -0,0 +1,311 @@
+import {AbstractValue, Value, VariablePointerValue, ListValue} from './Value';
+import {VariableAssignment} from './VariableAssignment';
+import {InkObject} from './Object';
+import {ListDefinitionsOrigin} from './ListDefinitionsOrigin';
+import {StoryException} from './StoryException';
+import {JsonSerialisation} from './JsonSerialisation';
+import {asOrThrows, asOrNull} from './TypeAssertion';
+import {tryGetValueFromMap} from './TryGetResult';
+import {throwNullException} from './NullException';
+
+export class VariablesState{
+	// The way variableChangedEvent is a bit different than the reference implementation.
+	// Originally it uses the C# += operator to add delegates, but in js we need to maintain
+	// an actual collection of delegates (ie. callbacks) to register a new one, there is a
+	// special ObserveVariableChange method below.
+	public variableChangedEventCallbacks: Array<(variableName: string, newValue: InkObject) => void> = [];
+	public variableChangedEvent(variableName: string, newValue: InkObject): void {
+		this.variableChangedEventCallbacks.forEach((callback) => {
+			callback(variableName, newValue);
+		});
+	}
+
+	get batchObservingVariableChanges(){
+		return this._batchObservingVariableChanges;
+	}
+	set batchObservingVariableChanges(value: boolean){
+		value = !!value; // TODO: Might have become useless with TypeScript.
+		this._batchObservingVariableChanges = value;
+		if (value) {
+			this._changedVariables = new Set();
+		}
+
+		else {
+			if (this._changedVariables != null) {
+				for (let variableName of this._changedVariables) {
+					let currentValue = this._globalVariables.get(variableName);
+					if (!currentValue) {
+						throwNullException('currentValue');
+					} else {
+						this.variableChangedEvent(variableName, currentValue);
+					}
+				}
+			}
+		}
+	}
+
+	get callStack(){
+		return this._callStack;
+	}
+	set callStack(callStack: any /* CallStack */){
```

Comment:
Why can't we use the `CallStack` type here ?

Reply:
Definitely, I guess that Callstack wasn't merged yet when the PR was first worked on.